app/uk/gov/hmrc/exports/migrations/ExportsMigrationTool.scala

```diff
@@ -0,0 +1,110 @@
+import play.api.Logger
+import uk.gov.hmrc.exports.migrations.changelogs.MigrationDefinition
+import uk.gov.hmrc.exports.migrations.exceptions.{ExportsMigrationException, LockCheckException}
+
+object ExportsMigrationTool {
+
+  private val lockCollectionName = "exportsMigrationLock"
+  private val changeEntryCollectionName = "exportsMigrationChangeLog"
+
+  def apply(mongoDatabase: MongoDatabase): ExportsMigrationTool = {
+    val migrationsRegistry = new MigrationsRegistry
+    ExportsMigrationTool(mongoDatabase, migrationsRegistry)
+  }
+
+  def apply(mongoDatabase: MongoDatabase, migrationsRegistry: MigrationsRegistry): ExportsMigrationTool = {
+    val lockRepository = new LockRepository(lockCollectionName, mongoDatabase)
+    val lockManager = new LockManager(lockRepository, new TimeUtils)
+    val changeEntryRepository = new ChangeEntryRepository(changeEntryCollectionName, mongoDatabase)
+
+    new ExportsMigrationTool(lockManager, migrationsRegistry, changeEntryRepository, mongoDatabase)
+  }
+}
+
+class ExportsMigrationTool(
+  val lockManager: LockManager,
+  val migrationsRegistry: MigrationsRegistry,
+  val changeEntryRepository: ChangeEntryRepository,
+  val database: MongoDatabase,
+  val throwExceptionIfCannotObtainLock: Boolean = false
+) {
+
+  private val logger = Logger(this.getClass)
+
+  private var enabled: Boolean = true
+
+  def execute(): Unit =
+    if (!isEnabled) {
+      logger.info("ExportsMigrationTool is disabled. Exiting.")
+    } else
+      try {
+        lockManager.acquireLockDefault()
+        executeMigration()
+      } catch {
+        case lockEx: LockCheckException =>
+          if (throwExceptionIfCannotObtainLock) {
+            logger.error(lockEx.getMessage)
+            throw new ExportsMigrationException(lockEx.getMessage)
+          }
+          logger.warn(lockEx.getMessage)
+          logger.warn("ExportsMigrationTool did not acquire process lock. EXITING WITHOUT RUNNING DATA MIGRATION")
+      } finally {
+        lockManager.releaseLockDefault() //we do it anyway, it's idempotent
+
+        logger.info("ExportsMigrationTool has finished his job.")
+      }
+
+  def isEnabled: Boolean = enabled
+  private[migrations] def setEnabled(enabled: Boolean): Unit =
+    this.enabled = enabled
+
+  private def executeMigration(): Unit = {
+    logger.info("ExportsMigrationTool starting the data migration sequence..")
+
+    migrationsRegistry.migrations.sorted.foreach(executeIfNewOrRunAlways)
+  }
+
+  private def executeIfNewOrRunAlways(migrDefinition: MigrationDefinition): Unit = {
+    val changeEntry = ChangeEntry(migrDefinition.migrationInformation)
+
+    try {
+      if (changeEntryRepository.isNewChange(changeEntry)) {
+        migrDefinition.migrationFunction(database)
+        changeEntryRepository.save(changeEntry)
+        logger.info(s"${changeEntry} applied")
+
+      } else if (isRunAlways(migrDefinition)) {
+        migrDefinition.migrationFunction(database)
```

Comment:
Minor point: Should this 'else if' block not be combined with the 'if'?   I.e. if new or run-always > run > update changeEntry db.   Otherwise (a) you have no record of when a run-always task last run and (b) if you switch run-always off it will run one more time as it will be "new"

Reply:
I did not spend more time refactoring. This is almost 1-to-1 copy. Only changed that to scala from java.

Reply:
That's fine.   I thought the "run always" was your feature.  Didn't realise that Mongock did that.